components/file_system/src/lib.rs

```diff
@@ -1,15 +1,147 @@
+    Write,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum IOType {
+    Other,
+    Read,
+    Write,
+    Coprocessor,
+    Flush,
+    Compaction,
+    Replication,
+    LoadBalance,
+    Import,
+    Export,
+}
+
+thread_local! {
+    static IO_TYPE: Cell<IOType> = Cell::new(IOType::Other)
+}
+
+fn set_io_type(new_io_type: IOType) {
+    IO_TYPE.with(|io_type| {
+        io_type.set(new_io_type);
+    });
+}
+
+fn get_io_type() -> IOType {
+    IO_TYPE.with(|io_type| io_type.get())
+}
+
+pub struct WithIOType {
+    previous_io_type: IOType,
+}
+
+impl WithIOType {
+    pub fn new(new_io_type: IOType) -> WithIOType {
+        let previous_io_type = get_io_type();
+        set_io_type(new_io_type);
+        WithIOType { previous_io_type }
+    }
+}
+
+impl Drop for WithIOType {
+    fn drop(&mut self) {
+        set_io_type(self.previous_io_type);
+    }
+}
+
+/// Indicates how large a buffer to pre-allocate before reading the entire file.
+fn initial_buffer_size(file: &File) -> usize {
+    // Allocate one extra byte so the buffer doesn't need to grow before the
+    // final `read` call at the end of the file.  Don't worry about `usize`
+    // overflow because reading will fail regardless in that case.
+    file.metadata().map(|m| m.len() as usize + 1).unwrap_or(0)
+}
+
+/// Write a slice as the entire contents of a file.
+pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> io::Result<()> {
+    File::create(path)?.write_all(contents.as_ref())
+}
+
+/// Read the entire contents of a file into a bytes vector.
+pub fn read<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> {
+    let mut file = File::open(path)?;
+    let mut bytes = Vec::with_capacity(initial_buffer_size(&file));
+    file.read_to_end(&mut bytes)?;
+    Ok(bytes)
+}
+
+/// Read the entire contents of a file into a string.
+pub fn read_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
+    let mut file = File::open(path)?;
+    let mut string = String::with_capacity(initial_buffer_size(&file));
+    file.read_to_string(&mut string)?;
+    Ok(string)
+}
+
+fn copy_imp(from: &Path, to: &Path, sync: bool) -> io::Result<u64> {
+    if !from.is_file() {
+        return Err(io::Error::new(
+            ErrorKind::InvalidInput,
+            "the source path is not an existing regular file",
+        ));
+    }
+
+    let mut reader = File::open(from)?;
+    let mut writer = File::create(to)?;
+    let perm = reader.metadata()?.permissions();
+
+    let ret = io::copy(&mut reader, &mut writer)?;
+    set_permissions(to, perm)?;
+    if sync {
+        writer.sync_all()?;
+        if let Some(parent) = to.parent() {
+            sync_dir(parent)?;
+        }
+    }
+    Ok(ret)
+}
+
+/// Copies the contents of one file to another. This function will also
+/// copy the permission bits of the original file to the destination file.
+pub fn copy<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> io::Result<u64> {
+    copy_imp(from.as_ref(), to.as_ref(), false)
```

Comment:
nit: comment what the bool is, or even better, use an enum
```suggestion
    copy_imp(from.as_ref(), to.as_ref(), false/*sync*/)
```